[ESQL] Support first/last functions.

mouhc1ne opened the pull request:
- Change `AggregatorImplementor` to use `Argument` instead of `AggregationParameter`.
- Compiling esql module resulted in only 2 modified `SpatialExtent*` classes

mouhc1ne committed:
[ESQL] Support first/last functions.
- Change `AggregatorImplementor` to use `Argument` instead of `AggregationParameter`.
- Compiling esql module resulted in only 2 modified `SpatialExtent*` classes

ivancea reviewed (approved):
LGTM!

craigtaverner commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/argument/Argument.java:
```diff
@@ -144,6 +186,8 @@ static Argument fromParameter(javax.lang.model.util.Types types, VariableElement
      */
     void read(MethodSpec.Builder builder, boolean blockStyle);
 
+    void read(MethodSpec.Builder builder, String accessor, String firstParam);
```
Probably needs javadoc like the adjacent methods.

craigtaverner replied:
Also, since all but one implementation is an empty implementation, we could use a `default` implementation here in the interface?

craigtaverner commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/argument/ArrayArgument.java:
```diff
@@ -135,14 +135,19 @@ public void read(MethodSpec.Builder builder, boolean blockStyle) {
     }
 
+    @Override
+    public void read(MethodSpec.Builder builder, String accessor, String firstParam) {
```
This could be deleted if we make a default method in the interface.

craigtaverner commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/argument/StandardArgument.java:
```diff
@@ -80,9 +80,19 @@ public void closeEvalToBlock(MethodSpec.Builder builder) {
+        if (invokeBlockEval.length == 1) {
+            builder.addStatement(invokeBlockEval[0]);
+            builder.endControlFlow();
+        } else {
```
I agree. Just have the for loop code, and skip the `length==1` check.

craigtaverner commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java:
```diff
@@ -334,33 +332,36 @@ private MethodSpec addRawInputExploded(boolean hasMask) {
-            builder.addStatement("$T $L = $L.asVector()", vectorType(p.type()), p.vectorName(), p.blockName());
-            builder.beginControlFlow("if ($L == null)", p.vectorName());
+
+        boolean isBlockArgument = aggParams.getFirst() instanceof BlockArgument;
```
If I remember correctly, we can vector optimization if all the values in the segment are single-valued. And the evaluators that were changed here were spatial extent aggregation over geo_point and cartesian point, which are almost always single-valued in the data-sets I've worked with.

craigtaverner reviewed (changes requested):
I did a quick review, and it seems to be like something is not quite right. It looks like the vector optimization for spatial points has been disabled.

mouhc1ne committed:
code reviews
- Move the `read(...)` method to `Argument` with default implementation, and remove all no-op implementations in child classes.
- Use `isBytesRef()` and `scratchName()`.

mouhc1ne committed:
code reviews

craigtaverner reviewed (approved):
OK. I see the optimization I was worried about did not even exist.

nik9000 commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/argument/PositionArgument.java:
```diff
@@ -17,6 +17,16 @@
 public record PositionArgument() implements Argument {
+    @Override
+    public String name() {
+        return null;
```
I suppose it does actually have a name, right?

nik9000 commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java:
```diff
@@ -334,33 +332,36 @@ private MethodSpec addRawInputExploded(boolean hasMask) {
-                    + aggParams.stream().map(AggregationParameter::blockName).collect(joining(", "))
-                    + (hasMask ? ", mask" : "")
-                    + ")"
+                "addRawVector(" + aggParams.stream().map(a -> a.vectorName()).collect(joining(", ")) + (hasMask ? ", mask" : "") + ")"
```
For this I think you want "if *all* arguments are BlockArgument` then call `addRawBlock` - otherwise you keep the `addRawVector`. I don't think it changes the generated code any, but feels a little more clear around the noop above.

nik9000 commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java:
```diff
@@ -334,33 +332,36 @@ private MethodSpec addRawInputExploded(boolean hasMask) {
         MethodSpec.Builder builder = initAddRaw(true, masked);
-        if (aggParams.getFirst().isArray()) {
+        if (aggParams.getFirst() instanceof BlockArgument) {
             builder.addComment("This type does not support vectors because all values are multi-valued");
```
@craigtaverner I *thought* you wrote this one originally. It's linked into the SpatialExtent aggs.

@mouhc1ne, I think if *all* of the arguments are `BlockArgument` then we can skip calling this method entirely. And just never make these methods. We never call them anyway.

nik9000 commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java:
```diff
@@ -427,14 +428,14 @@ private MethodSpec addRawBlock(boolean masked) {
             }
 
-            if (aggParams.getFirst().isArray()) {
+            if (aggParams.getFirst() instanceof BlockArgument) {
```
I'm not super happy we have this branch at all, but now is probably not the time to nuke it.

mouhc1ne committed:
code review
- Pass the `type` and `name` down to `PositionArgument` when it's instantiated from `Argument`. The explicit declaration of `type()` and `name()` methods isn't needed thus they're removed.
- Add `elementType()` method to `Argument` with default implementation returning the type, except for `BlockArgument` where it returns the inner type of elements in the block.
- Add `declareProcessParameter()` to `Argument` to simplify some logic in `AggregatorImplementer`.
- Do not call `addRawVector` methods for block types. This affect 2 `SpatialExtent*` classes where these methods are removed.

mouhc1ne committed:
code review
- Pass the `type` and `name` down to `PositionArgument` when it's instantiated from `Argument`. The explicit declaration of `type()` and `name()` methods isn't needed thus they're removed.
- Add `elementType()` method to `Argument` with default implementation returning the type, except for `BlockArgument` where it returns the inner type of elements in the block.
- Add `declareProcessParameter()` to `Argument` to simplify some logic in `AggregatorImplementer`.
- Do not call `addRawVector` methods for block types. This affect 2 `SpatialExtent*` classes where these methods are removed.

mouhc1ne committed:
code review

Commit:
[CI] Auto commit changes from spotless
